File: app/src/main/java/io/stewlab/stewdroid/StewDroid.java
package io.stewlab.stewdroid;

import android.app.Activity;
import android.content.Intent;
import android.util.ArraySet;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.godotengine.godot.Godot;
import org.godotengine.godot.plugin.GodotPlugin;
import org.godotengine.godot.plugin.SignalInfo;
import org.godotengine.godot.plugin.UsedByGodot;

import java.util.List;
import java.util.Set;

import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;


// StewDroid.gdap below

//[config]
//
//name="StewDroid"
//binary_type="local"
//binary="app-debug.aar"
//
//[dependencies]
//
//local=[]
//remote=[]
//custom_maven_repos=[]


public class StewDroid extends GodotPlugin {

    public StewDroid(Godot godot) {
        super(godot);
    }

    @UsedByGodot
    public String getHello() {
        return "Hello from StewDroid!";
    }

    @UsedByGodot
    public void getHelloSignal(String string) {
        emitSignal("testSignal", string);
    }

    @NonNull
    @Override
    public Set<SignalInfo> getPluginSignals() {
        Set<SignalInfo> signals = new ArraySet<>();
        signals.add(new SignalInfo("testSignal", String.class));
        return signals;
    }

    @NonNull
    @Override
    public String getPluginName() {
        return "StewDroid";
    }

    @Override
    protected Godot getGodot() {
        return super.getGodot();
    }

    @Nullable
    @Override
    protected Activity getActivity() {
        return super.getActivity();
    }

    @Nullable
    @Override
    public View onMainCreate(Activity activity) {
        return super.onMainCreate(activity);
    }

    @Override
    public void onMainActivityResult(int requestCode, int resultCode, Intent data) {
        super.onMainActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onMainRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onMainRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    public void onMainPause() {
        super.onMainPause();
    }

    @Override
    public void onMainResume() {
        super.onMainResume();
    }

    @Override
    public void onMainDestroy() {
        super.onMainDestroy();
    }

    @Override
    public boolean onMainBackPressed() {
        return super.onMainBackPressed();
    }

    @Override
    public void onGodotSetupCompleted() {
        super.onGodotSetupCompleted();
    }

    @Override
    public void onGodotMainLoopStarted() {
        super.onGodotMainLoopStarted();
    }

    @Override
    public void onGLDrawFrame(GL10 gl) {
        super.onGLDrawFrame(gl);
    }

    @Override
    public void onGLSurfaceChanged(GL10 gl, int width, int height) {
        super.onGLSurfaceChanged(gl, width, height);
    }

    @Override
    public void onGLSurfaceCreated(GL10 gl, EGLConfig config) {
        super.onGLSurfaceCreated(gl, config);
    }

//    @NonNull
//    @Override
//    public List<String> getPluginMethods() {
//        return super.getPluginMethods();
//    }

    @NonNull
    @Override
    protected Set<String> getPluginGDNativeLibrariesPaths() {
        return super.getPluginGDNativeLibrariesPaths();
    }

    @Override
    public boolean shouldBeOnTop() {
        return super.shouldBeOnTop();
    }

    @Override
    protected void runOnUiThread(Runnable action) {
        super.runOnUiThread(action);
    }

    @Override
    protected void runOnRenderThread(Runnable action) {
        super.runOnRenderThread(action);
    }

    @Override
    protected void emitSignal(String signalName, Object... signalArgs) {
        super.emitSignal(signalName, signalArgs);
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        return super.equals(obj);
    }

    @NonNull
    @Override
    protected Object clone() throws CloneNotSupportedException {
        return super.clone();
    }

    @NonNull
    @Override
    public String toString() {
        return super.toString();
    }

    @Override
    protected void finalize() throws Throwable {
        super.finalize();
    }
}

File: app/src/main/java/io/stewlab/stewdroid/StewDroid.kt
package io.stewlab.stewdroid

import android.app.Activity
import android.content.Intent
import android.util.ArraySet
import android.view.View
import androidx.fragment.app.FragmentActivity
import io.stewlab.stewdroid.admob.AdMob
import io.stewlab.stewdroid.admob.RewardedAdListener
import io.stewlab.stewdroid.enum.AdMobSignals
import io.stewlab.stewdroid.enum.Signals
import org.godotengine.godot.Godot
import org.godotengine.godot.plugin.GodotPlugin
import org.godotengine.godot.plugin.SignalInfo
import org.godotengine.godot.plugin.UsedByGodot
import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10

class StewDroid(godot: Godot?) : GodotPlugin(godot) {

    private val godotActivity: FragmentActivity? = godot?.activity
    private val adMob: AdMob = AdMob()

    init {

    }

    // AdMob Functions
    @UsedByGodot
    fun initializeAdmob() {
        adMob.initialize(godotActivity)
    }

    @UsedByGodot
    fun loadRewardedAd() {
        adMob.loadRewardedAd(godotActivity, object: RewardedAdListener {
            override fun onAdLoaded() {
                emitSignal(AdMobSignals.REWARDED_AD_LOADED.signalName, "")
            }

            override fun onAdFailedToLoad(errorCode: Int) {
                emitSignal(AdMobSignals.REWARDED_AD_FAILED_TO_LOAD.signalName, "")
            }

            override fun onAdShowedFullScreenContent() {
                TODO("Not yet implemented")
            }

            override fun onAdFailedToShowFullScreenContent() {
                TODO("Not yet implemented")
            }

            override fun onAdDismissedFullScreenContent() {
                TODO("Not yet implemented")
            }

            override fun onAdClicked() {
                TODO("Not yet implemented")
            }

            override fun onAdImpression() {
                TODO("Not yet implemented")
            }

            override fun onUserEarnedReward(type: String, amount: Int) {
                emitSignal(AdMobSignals.REWARDED_AD_USER_EARNED_REWARD.signalName, "")
            }
        })
    }

    @UsedByGodot
    fun showRewardAd() {
        adMob.showRewardAd(godotActivity)
    }

    @UsedByGodot
    fun shareText(text : String) {

        val sendIntent: Intent = Intent().apply {
            action = Intent.ACTION_SEND
            putExtra(Intent.EXTRA_TEXT, text)
            type = "text/plain"
        }

        val shareIntent = Intent.createChooser(sendIntent, null)

        godotActivity?.startActivity(shareIntent)

    }

    @UsedByGodot
    fun getHello(): String {
        return "Hello from Kotlin StewDroid!"
    }

    override fun getPluginName(): String {
        return "StewDroid"
    }

    override fun getPluginSignals(): MutableSet<SignalInfo> {

        val signals = ArraySet<SignalInfo>()

        signals.add(SignalInfo(
            Signals.STEWDROID_INIT.signalName,
            String::class.java
        ))

        signals.add(SignalInfo(
            AdMobSignals.ADMOB_INIT.signalName,
            String::class.java
        ))

        signals.add(SignalInfo(
            AdMobSignals.REWARDED_AD_LOADED.signalName,
            String::class.java
        ))

        signals.add(SignalInfo(
            AdMobSignals.REWARDED_AD_FAILED_TO_LOAD.signalName,
            String::class.java
        ))

        signals.add(SignalInfo(
            AdMobSignals.REWARDED_AD_USER_EARNED_REWARD.signalName,
            String::class.java
        ))

        return signals
    }

    override fun getGodot(): Godot {
        return super.getGodot()
    }

    override fun getActivity(): Activity? {
        return super.getActivity()
    }

    override fun onMainCreate(activity: Activity?): View? {
        return super.onMainCreate(activity)
    }

    override fun onMainActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onMainActivityResult(requestCode, resultCode, data)
    }

    override fun onMainRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>?,
        grantResults: IntArray?
    ) {
        super.onMainRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onMainPause() {
        super.onMainPause()
    }

    override fun onMainResume() {
        super.onMainResume()
    }

    override fun onMainDestroy() {
        super.onMainDestroy()
    }

    override fun onMainBackPressed(): Boolean {
        return super.onMainBackPressed()
    }

    override fun onGodotSetupCompleted() {
        super.onGodotSetupCompleted()
        emitSignal(Signals.STEWDROID_INIT.signalName, "StewDroid has been initialized!")
    }

    override fun onGodotMainLoopStarted() {
        super.onGodotMainLoopStarted()
    }

    override fun onGLDrawFrame(gl: GL10?) {
        super.onGLDrawFrame(gl)
    }

    override fun onGLSurfaceChanged(gl: GL10?, width: Int, height: Int) {
        super.onGLSurfaceChanged(gl, width, height)
    }

    override fun onGLSurfaceCreated(gl: GL10?, config: EGLConfig?) {
        super.onGLSurfaceCreated(gl, config)
    }

//    override fun getPluginMethods(): MutableList<String> {
//        return super.getPluginMethods()
//    }

    override fun getPluginGDNativeLibrariesPaths(): MutableSet<String> {
        return super.getPluginGDNativeLibrariesPaths()
    }

    override fun shouldBeOnTop(): Boolean {
        return super.shouldBeOnTop()
    }

    override fun runOnUiThread(action: Runnable?) {
        super.runOnUiThread(action)
    }

    override fun runOnRenderThread(action: Runnable?) {
        super.runOnRenderThread(action)
    }

    override fun emitSignal(signalName: String?, vararg signalArgs: Any?) {
        super.emitSignal(signalName, *signalArgs)
    }
}